### PR TITLE
feat: Add optimized diff routine

### PR DESCRIPTION
**Pull request by taylorsw04**
This change adds a method to BTree that computes the difference between two trees. It is optimized for the case of trees created via the clone/with APIs, skipping shared nodes in these cases. Microbenchmarks reflect the expected asymptotic change, showing huge speedups in the case of mostly shared nodes.

### CODE REVIEW

**Comment by taylorsw04** [comment]
@qwertie I'm hoping to use this in production soon, and I'm wondering if you'll have a chance to take a look at it in the near future. If not, I can always publish a package from my fork, but I'd prefer a unified approach :) Thanks!

**Comment by qwertie** [comment]
I'm sorry, I've been busy and overlooked this. Did you prepare for production already?

**Comment by taylorsw04** [comment]
Ya, I created [my own package](https://www.npmjs.com/package/@taylorsw04/sorted-btree) as a workaround, but we'd certainly prefer to remove that and pull from this package :)

**Comment by qwertie** [comment]
I wrote those comments a week or two ago but I didn't realize that apparently they aren't public until I click "submit review"?

Anyways, diffing is a cool and unique feature, and I'd like to thank you guys for working on that!

File: readme.md [hunk 1]
```diff
@@ -1,422 +1,4 @@
...
-
-### Endnote ###
-
-♥ This package was made to help people [learn TypeScript & React](http://typescript-react-primer.loyc.net/).
-
-Are you a C# developer? You might like the similar data structures I made for C#: 
-BDictionary, BList, etc. See http://core.loyc.net/collections/
-
-You might think that the package name "sorted btree" is overly redundant, but I _did_ make a data structure similar to B+ Tree that is _not_ sorted. I called it the [A-List](http://core.loyc.net/collections/alists-part1) (C#). But yeah, the names `btree` and `bplustree` were already taken, so what was I supposed to do, right?
+This is a temporary fork of the [B+ tree library](https://github.com/qwertie/btree-typescript) that adds an optimized diff method.
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply readme.md:4]: Whoops, just pushed the revert.

File: b+tree.ts [hunk 2]
```diff
@@ -489,129 +489,157 @@ export default class BTree<K=any, V=any> implements ISortedMapF<K,V>, ISortedMap
...
-        const { leaf: leafSrc, internalSpine: internalSpineSrc, levelIndices: levelIndicesSrc } = src;
-        const { leaf: leafDst, internalSpine: internalSpineDst, levelIndices: levelIndicesDst } = dst;
-        if (leafSrc || leafDst) {
-          if (leafSrc && leafDst && cursorOrder === 0) {
+  diff<R>(
+    other: BTree<K, V>,
+    onlyThis?: (k: K, v: V) => { break?: R } | void,
+    onlyOther?: (k: K, v: V) => { break?: R } | void,
+    different?: (k: K, vThis: V, vOther: V) => { break?: R} | void
+  ): R | undefined {
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply b+tree.ts:586]: How about `diffAgainst`?

**Comment by taylorsw04** [comment]
Thank you for the responsiveness and for making an awesome library!

File: b+tree.ts [hunk 3]
```diff
@@ -489,129 +489,157 @@ export default class BTree<K=any, V=any> implements ISortedMapF<K,V>, ISortedMap
...
+   * Computes the differences between `this` and `other`.
+   * For efficiency, the diff is returned via invocations of supplied handlers.
+   * The computation is optimized for the case in which the two trees have large amounts 
+   * of shared data (obtained by calling the `clone` or `with` APIs) and will avoid 
+   * any iteration of shared state.
+   * The handlers can cause computation to early exit by returning {break: R}.
+   * @param other The tree to compute a diff against.
+   * @param onlyThis Callback invoked for all keys only present in `this`.
+   * @param onlyOther Callback invoked for all keys only present in `other`.
+   * @param different Callback invoked for all keys with differing values.
```
[the remark this answers is not in the document]
Comment (taylorsw04) [reply b+tree.ts:579]: Done.

File: b+tree.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (taylorsw04) [reply b+tree.ts:586]: I went ahead and renamed it to `diffAgainst`, but am happy with `diffWith` if you'd prefer.

**Pull request merged**